bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieStatus.java

```diff
@@ -0,0 +1,246 @@
+     *
+     * @param file file that you want to write to
+     * @param body content to write
+     * @throws IOException
+     */
+    private void writeToFile(File file, String body) throws IOException {
+        FileOutputStream fos = new FileOutputStream(file);
+        BufferedWriter bw = null;
+        try {
+            bw = new BufferedWriter(new OutputStreamWriter(fos, UTF_8));
+            bw.write(body);
+        } finally {
+            if (bw != null) {
+                bw.close();
+            }
+            fos.close();
+        }
+    }
+
+    /**
+     * Read bookie status from the status files, and update the bookie status if read succeed.
+     * If a status file is not readable or not found, it will skip and try to read from the next file.
+     *
+     * @param directories list of directories that store the status file
+     * @return true if successfully read from file, otherwise false.
+     */
+    boolean readFromDirectories(List<File> directories) {
+        boolean success = false;
+        for (File dir : directories) {
+            File statusFile = new File(dir, BOOKIE_STATUS_FILENAME);
+            try {
+                BookieStatus status = readFromFile(statusFile);
+                if (null != status) {
+                    synchronized (status) {
+                        if (status.lastUpdateTime > this.lastUpdateTime) {
+                            this.lastUpdateTime = status.lastUpdateTime;
+                            this.layoutVersion = status.layoutVersion;
+                            this.bookieMode = status.bookieMode;
+                            success = true;
+                        }
+                    }
+                }
+            } catch (IOException e) {
+                LOG.warn("IOException while trying to read bookie status from directory {}." +
+                    " This is fine if not all directories failed.", dir);
+            } catch (IllegalArgumentException e ){
+                LOG.warn("IllegalArgumentException while trying to read bookie status from directory {}." +
+                    " This is fine if not all directories failed.", dir);
+            }
+        }
+        if (success) {
+            LOG.info("Successfully retrieve bookie status {} from disks.", getBookieMode());
+        } else {
+            LOG.warn("Failed to retrieve bookie status from disks." +
+                    " Fall back to current or default bookie status: {}", getBookieMode());
+        }
+        return success;
+    }
+
+
+    /**
+     * Function to read the bookie status from a single file
+     *
+     * @param file file to read from
+     * @return BookieStatus if not error, null if file not exist or any exception happens
+     * @throws IOException
+     */
+    private BookieStatus readFromFile(File file)
+            throws IOException, IllegalArgumentException {
+        if (!file.exists()) {
+            return null;
```

Comment:
this is not an exception, because if there is no file, it means there is not status stored on disks.